### ISSUE TITLE
Unable to execute child process!

### ISSUE THREAD

**Issue by noodlehair-git** [oscarlab/graphene#1989]
I am trying to run a python script that launches Tor, with Graphene (below).

``` 
  import io
  import pycurl
  import time
  import stem.process
  from stem.util import term
  import threading
  import os


  def mtor(): 

  SOCKS_PORT = 7000

  ctr_old = 0
  ctr_new = 5004445
  threads = list()
  name = 0
  global ctr
  ctr = 0

  def query(url,range, name):
    """
    Uses pycurl to fetch a site using the proxy on the SOCKS_PORT.
    """
    global ctr
    output = io.BytesIO()

    query = pycurl.Curl()
    query.setopt(pycurl.URL, url)
    query.setopt(pycurl.PROXY, 'localhost')
    query.setopt(pycurl.PROXYPORT, SOCKS_PORT)
    query.setopt(pycurl.PROXYTYPE, pycurl.PROXYTYPE_SOCKS5_HOSTNAME)
    query.setopt(pycurl.RANGE, range)
    query.setopt(pycurl.WRITEFUNCTION, output.write)

    try:
      query.perform()
      buffer = output.getvalue()
      while ctr != name:
        continue
      print("ctr:"+str(ctr))
      ctr = ctr + 1
      
      if b"416 Requested Range Not Satisfiable" in buffer:
        print("Download Complete")
        print(time.time() - start_time)
        tor_process.kill()
        os._exit(1)
      else:
        print(buffer)
      del buffer
    except pycurl.error as exc:
      raise ValueError("Unable to reach %s (%s)" % (url, exc))
      del output

    except pycurl.error as exc:
      return "Unable to reach %s (%s)" % (url, exc)

  tor_process = stem.process.launch_tor_with_config(
    config = {
      'SocksPort': str(SOCKS_PORT),
      'MaxCircuitDirtiness': str(1),
    }
  )
  print("LAUNCHED TOR")

  start_time = time.time()
  while True:
    range = str(ctr_old)+"-"+str(ctr_new)
    x = threading.Thread(target=query, args=('any_website.com', range, name,))
    
    threads.append(x)
    x.start()
    
    time.sleep(2)
    ctr_old = ctr_new + 1
    ctr_new = ctr_old + 5004445
    name = name + 1

    

    `

```

Initially, I was using the same ```python.manifest.template``` as given in the Graphene-Examples but then I got the following error 

```
File "/usr/lib/python3/dist-packages/stem/process.py", line 285, in launch_tor_with_config
    return launch_tor(tor_cmd, args, torrc_path, completion_percent, init_msg_handler, timeout, take_ownership)
  File "/usr/lib/python3/dist-packages/stem/process.py", line 123, in launch_tor
    tor_process = subprocess.Popen(runtime_args, stdout = subprocess.PIPE, stdin = subprocess.PIPE, stderr = subprocess.PIPE)
  File "/usr/lib/python3.6/subprocess.py", line 729, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.6/subprocess.py", line 1295, in _execute_child
    restore_signals, start_new_session, preexec_fn)
PermissionError: [Errno 13] Permission denied

```

And realized that the script needs a Child process permissions. So I referred to ```python->sh->python``` example but I guess I am doing something wrong because I get the following error again:

```
Not an allowed enclave (mr_enclave = dc3be08ad3fff9e9e2275063ee20f6ec3b32f7758d69a3fbdf31e3114d8ce47b). Maybe missing 'sgx.trusted_children' in the manifest file?
Failed to receive local report via RPC: -20
Failed to initialize child process: -20
Not an allowed enclave (mr_enclave = dc3be08ad3fff9e9e2275063ee20f6ec3b32f7758d69a3fbdf31e3114d8ce47b). Maybe missing 'sgx.trusted_children' in the manifest file?
Failed to receive local report via RPC: -20
Failed to initialize child process: -20
Not an allowed enclave (mr_enclave = dc3be08ad3fff9e9e2275063ee20f6ec3b32f7758d69a3fbdf31e3114d8ce47b). Maybe missing 'sgx.trusted_children' in the manifest file?
Failed to receive local report via RPC: -20
Failed to initialize child process: -20

File "/usr/lib/python3/dist-packages/stem/process.py", line 285, in launch_tor_with_config
    return launch_tor(tor_cmd, args, torrc_path, completion_percent, init_msg_handler, timeout, take_ownership)
  File "/usr/lib/python3/dist-packages/stem/process.py", line 123, in launch_tor
    tor_process = subprocess.Popen(runtime_args, stdout = subprocess.PIPE, stdin = subprocess.PIPE, stderr = subprocess.PIPE)
  File "/usr/lib/python3.6/subprocess.py", line 729, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.6/subprocess.py", line 1295, in _execute_child
    restore_signals, start_new_session, preexec_fn)
PermissionError: [Errno 13] Permission denied

```

My manifest files are as follows:

[Manifests.zip](https://github.com/oscarlab/graphene/files/5618538/Manifests.zip)

Basically, my issue is that I am unable to launch Tor with Graphene.

**Comment by noodlehair-git** [comment]
Update: I also gave ```python -> tor -> python``` a try and that did not work either:(

**Comment by dimakuv** [comment]
Why are you using `python -> sh -> python` manifests? From the Python script, it looks like your application creates one Curl (`curl`) child process and one Tor child process.

To understand what your original native application does, run it under `strace -ff [native-python-run]` and carefully examine the output strace log. In particular, search for `execve()` system calls -- these are the ones that indicate spawned child processes. I'm pretty sure you'll see something like `execve(curl)` and `execve(tor)` there. So you'll need to create manifests for those executables.

**Comment by noodlehair-git** [comment]
In that case, I should be trying ```python -> curl + tor -> python``` manifest path?

**Comment by dimakuv** [comment]
You should start with understanding what your original application does. The best of doing this is `strace -ff`.

**Comment by noodlehair-git** [comment]
Thanks for your response!

Here is my trace file (for a successful script run). 
[trace.txt](https://github.com/oscarlab/graphene/files/5610823/trace.txt)

This is the only ```execve()``` instance I found. 

```execve("/usr/bin/python3", ["python3", "scripts/dummy-web-server.py"], 0x7ffe76e2b388 /* 21 vars */) = 0```

**Comment by noodlehair-git** [comment]
I referred to ``` launch_tor_with_config``` and saw that it creates a temporary file (https://stem.torproject.org/_modules/stem/process.html#launch_tor_with_config). I added those permissions for ```torrc-``` in the manifest and that did not help either:(

**Comment by dimakuv** [comment]
I don't think you did `strace -ff`. Notice the `-ff` flag -- this is to trace not only the main parent, but also all children.

**Comment by noodlehair-git** [comment]
I guess you're right. I forgot to attach a bunch of other trace files generated to track the children. So here's what I found:
```
execve("/usr/bin/python3", ["python3", "dummy-web-server.py"], 0x7ffe5c2cce50 /* 21 vars */) = 0
execve("/home/arushiarora/bin/tor", ["tor", "--version"], 0x7ffd08220740 /* 21 vars */) = -1 ENOENT (No such file or directory)
execve("/home/arushiarora/.local/bin/tor", ["tor", "--version"], 0x7ffd08220740 /* 21 vars */) = -1 ENOENT (No such file or directory)
execve("/usr/local/sbin/tor", ["tor", "--version"], 0x7ffd08220740 /* 21 vars */) = -1 ENOENT (No such file or directory)
execve("/usr/local/bin/tor", ["tor", "--version"], 0x7ffd08220740 /* 21 vars */) = 0
execve("/bin/sh", ["/bin/sh", "-c", "uname -p 2> /dev/null"], 0x7ffd08220740 /* 21 vars */) = 0
execve("/bin/uname", ["uname", "-p"], 0x5598a7f9ebe0 /* 21 vars */) = 0
brk(NULL)                               
```

Also here are the trace files for your reference:


[trace.zip](https://github.com/oscarlab/graphene/files/5618451/trace.zip)

**Comment by dimakuv** [comment]
You'll need to add the manifests for `/usr/local/bin/tor` and `/bin/sh` and `/bin/uname`... Alternatively, you could wait for a week or two until Graphene's multi-process logic is significantly re-written (it is work in progress currently). In the new version of Graphene, this kind of multi-process manifests will go away, and it will be generally much easier to port.

**Comment by noodlehair-git** [comment]
Oh great! I will wait for the multi-process logic to be re-written. Hopefully I can contribute to Graphene, post that, with my Tor client examples:)

**Comment by mkow** [comment]
@noodlehair-git Could you try #2050?

**Comment by noodlehair-git** [comment]
So basically with the new update, ```python.manifest.template``` will be enough to run the script that I wanted to? @mkow

**Comment by mkow** [comment]
Yes. You just need to update it to the new format (see our documentation) and include all used files as trusted or allowed in the manifest.

**Comment by noodlehair-git** [comment]
Thanks. I did all the modifications to the manifest. 
[python.manifest.template.zip](https://github.com/oscarlab/graphene/files/5840094/python.manifest.template.zip)
But running the above script involves launching tor and gives the following error ```OSError: 'tor' isn't available on your system. Maybe it's not in your PATH?```. I guess it is not able to access the executable file (```/usr/local/bin/tor```). How should I mention this in the manifest?

**Comment by dimakuv** [comment]
The manifest file that you attached is still in the old format... Forgot to attach the newest manifest?

> But running the above script involves launching tor and gives the following error OSError: 'tor' isn't available on your system. Maybe it's not in your PATH?. I guess it is not able to access the executable file (/usr/local/bin/tor). How should I mention this in the manifest?

I see that you already mounted the `/usr/` directory, so the file `/usr/local/bin/tor` must be visible to Graphene.

I also see that your `PATH` environment variable is defined as follows in your manifest file:
```
loader.env.PATH = $(PYTHONHOME)/bin:/usr/bin:/bin
```

Notice how there is no `/usr/local/bin` entry. So maybe just append this entry to the end, like this:
```
loader.env.PATH = $(PYTHONHOME)/bin:/usr/bin:/bin:/usr/local/bin
```

Also, you of course need to declare this TOR file (`/usr/local/bin/tor`) as `sgx.trusted_files` in the manifest.

**Comment by noodlehair-git** [comment]
Thanks! 
Sorry, my bad. 
[python.manifest.template.zip](https://github.com/oscarlab/graphene/files/5845041/python.manifest.template.zip)

I made the changes suggested by you and its now able to find the path. Running Tor needs maybe permission to something else as well. 
Based on Tor start up message (that is displayed when you generally launch Tor), I gave access to the ```geoip``` database in the manifest. 

```
Jan 20 12:19:39.056 [notice] Tor 0.4.5.0-alpha-dev (git-3c884bc9e077e997) running on Linux with Libevent 2.1.8-stable, OpenSSL 1.1.1, Zlib 1.2.11, Liblzma 5.2.2, Libzstd N/A and Glibc 2.27 as libc.
Jan 20 12:19:39.056 [notice] Tor can't help you if you use it wrong! Learn how to be safe at https://www.torproject.org/download/download#warning
Jan 20 12:19:39.056 [notice] This version is not a stable Tor release. Expect more bugs than usual.
Jan 20 12:19:39.056 [notice] Read configuration file "/usr/local/etc/tor/torrc".
Jan 20 12:19:39.057 [warn] ControlPort is open, but no authentication method has been configured.  This means that any program on your computer can reconfigure your Tor.  That's bad!  You should upgrade your Tor controller as soon as possible.
Jan 20 12:19:39.057 [notice] Opening Socks listener on 127.0.0.1:9050
Jan 20 12:19:39.057 [notice] Opened Socks listener connection (ready) on 127.0.0.1:9050
Jan 20 12:19:39.057 [notice] Opening Control listener on 127.0.0.1:9051
Jan 20 12:19:39.057 [notice] Opened Control listener connection (ready) on 127.0.0.1:9051
Jan 20 12:19:39.000 [notice] Parsing GEOIP IPv4 file /usr/local/share/tor/geoip.
Jan 20 12:19:39.000 [notice] Parsing GEOIP IPv6 file /usr/local/share/tor/geoip6.
Jan 20 12:19:39.000 [notice] Bootstrapped 0% (starting): Starting
Jan 20 12:19:39.000 [notice] Starting with guard context "default"
Jan 20 12:19:40.000 [notice] Bootstrapped 5% (conn): Connecting to a relay
Jan 20 12:19:40.000 [notice] Bootstrapped 10% (conn_done): Connected to a relay
Jan 20 12:19:40.000 [notice] Bootstrapped 14% (handshake): Handshaking with a relay
Jan 20 12:19:41.000 [notice] Bootstrapped 15% (handshake_done): Handshake with a relay done
Jan 20 12:19:41.000 [notice] Bootstrapped 20% (onehop_create): Establishing an encrypted directory connection
Jan 20 12:19:41.000 [notice] Bootstrapped 25% (requesting_status): Asking for networkstatus consensus
Jan 20 12:19:41.000 [notice] Bootstrapped 30% (loading_status): Loading networkstatus consensus
Jan 20 12:19:44.000 [notice] Bootstrapped 45% (requesting_descriptors): Asking for relay descriptors
Jan 20 12:19:44.000 [notice] I learned some more directory information, but not enough to build a circuit: We need more microdescriptors: we have 11/6771, and can only build 0% of likely paths. (We have 0% of guards bw, 0% of midpoint bw, and 0% of exit bw = 0% of path bw.)
Jan 20 12:19:44.000 [notice] I learned some more directory information, but not enough to build a circuit: We need more microdescriptors: we have 11/6771, and can only build 0% of likely paths. (We have 0% of guards bw, 0% of midpoint bw, and 0% of exit bw = 0% of path bw.)
Jan 20 12:19:44.000 [notice] I learned some more directory information, but not enough to build a circuit: We need more microdescriptors: we have 11/6771, and can only build 0% of likely paths. (We have 0% of guards bw, 0% of midpoint bw, and 0% of exit bw = 0% of path bw.)
Jan 20 12:19:45.000 [notice] Bootstrapped 50% (loading_descriptors): Loading relay descriptors
Jan 20 12:19:47.000 [notice] Bootstrapped 56% (loading_descriptors): Loading relay descriptors
Jan 20 12:19:47.000 [notice] Bootstrapped 63% (loading_descriptors): Loading relay descriptors
Jan 20 12:19:49.000 [notice] Bootstrapped 68% (loading_descriptors): Loading relay descriptors
Jan 20 12:19:55.000 [notice] Bootstrapped 75% (enough_dirinfo): Loaded enough directory info to build circuits
Jan 20 12:19:56.000 [notice] Bootstrapped 80% (ap_conn): Connecting to a relay to build circuits
Jan 20 12:19:56.000 [notice] Bootstrapped 85% (ap_conn_done): Connected to a relay to build circuits
Jan 20 12:19:56.000 [notice] Bootstrapped 89% (ap_handshake): Finishing handshake with a relay to build circuits
Jan 20 12:19:57.000 [notice] Bootstrapped 90% (ap_handshake_done): Handshake finished with a relay to build circuits
Jan 20 12:19:57.000 [notice] Bootstrapped 95% (circuit_create): Establishing a Tor circuit
Jan 20 12:19:57.000 [notice] Bootstrapped 100% (done): Done
```
But it still throws this error

```
  File "/usr/lib/python3/dist-packages/stem/process.py", line 285, in launch_tor_with_config
    return launch_tor(tor_cmd, args, torrc_path, completion_percent, init_msg_handler, timeout, take_ownership)
  File "/usr/lib/python3/dist-packages/stem/process.py", line 153, in launch_tor
    raise OSError('Process terminated: %s' % last_problem)
OSError: Process terminated: Timed out
```

Based on the ```process.py``` file [here](https://programtalk.com/vs2/python/12155/spiderfoot/ext/stem/process.py/), I am still trying to figure out what more permissions does this script require. Thanks so much for all your help.

**Comment by dimakuv** [comment]
At this point, you need to enable the Graphene log in your manifest:
- `loader.debug_type = "inline"` for older versions of Graphene
- `loader.log_level = "all"` for newer versions of Graphene

Read more info here: https://graphene.readthedocs.io/en/latest/manifest-syntax.html#log-level

The Graphene log messages will hopefully give you more information on what goes wrong. Probably, Tor wants to open some file that doesn't exist inside the graphenized version, or Tor tries to use a system call that is not implemented in Graphene (in this case you'll see some `-38` errors).

**Comment by dimakuv** [comment]
No reply for 6 months, considering this issue closed.
